<h1 id="closures-anonymous-functions-that-can-capture-their-environment">Closures: Anonymous Functions that Can Capture Their Environment</h1> <p>Rust’s closures are anonymous functions you can save in a variable or pass as arguments to other functions. You can create the closure in one place and then call the closure to evaluate it in a different context. Unlike functions, closures can capture values from the scope in which they’re defined. We’ll demonstrate how these closure features allow for code reuse and behavior customization.</p> <h3 id="creating-an-abstraction-of-behavior-with-closures">Creating an Abstraction of Behavior with Closures</h3> <p>Let’s work on an example of a situation in which it’s useful to store a closure to be executed later. Along the way, we’ll talk about the syntax of closures, type inference, and traits.</p> <p>Consider this hypothetical situation: we work at a startup that’s making an app to generate custom exercise workout plans. The backend is written in Rust, and the algorithm that generates the workout plan takes into account many factors, such as the app user’s age, body mass index, exercise preferences, recent workouts, and an intensity number they specify. The actual algorithm used isn’t important in this example; what’s important is that this calculation takes a few seconds. We want to call this algorithm only when we need to and only call it once so we don’t make the user wait more than necessary.</p> <p>We’ll simulate calling this hypothetical algorithm with the function <code>simulated_expensive_calculation</code> shown in Listing 13-1, which will print <code>calculating slowly...</code>, wait for two seconds, and then return whatever number we passed in.</p> <p><span class="filename">Filename: src/main.rs</span></p> <pre data-language="rust">use std::thread;
use std::time::Duration;

fn simulated_expensive_calculation(intensity: u32) -&gt; u32 {
    println!("calculating slowly...");
    thread::sleep(Duration::from_secs(2));
    intensity
}

fn main() {}
</pre> <p><span class="caption">Listing 13-1: A function to stand in for a hypothetical calculation that takes about 2 seconds to run</span></p> <p>Next is the <code>main</code> function, which contains the parts of the workout app important for this example. This function represents the code that the app will call when a user asks for a workout plan. Because the interaction with the app’s frontend isn’t relevant to the use of closures, we’ll hardcode values representing inputs to our program and print the outputs.</p> <p>The required inputs are these:</p> <ul> <li>An intensity number from the user, which is specified when they request a workout to indicate whether they want a low-intensity workout or a high-intensity workout</li> <li>A random number that will generate some variety in the workout plans</li> </ul> <p>The output will be the recommended workout plan. Listing 13-2 shows the <code>main</code> function we’ll use.</p> <p><span class="filename">Filename: src/main.rs</span></p> <pre data-language="rust">use std::thread;
use std::time::Duration;

fn simulated_expensive_calculation(intensity: u32) -&gt; u32 {
    println!("calculating slowly...");
    thread::sleep(Duration::from_secs(2));
    intensity
}

fn generate_workout(intensity: u32, random_number: u32) {}

fn main() {
    let simulated_user_specified_value = 10;
    let simulated_random_number = 7;

    generate_workout(simulated_user_specified_value, simulated_random_number);
}
</pre> <p><span class="caption">Listing 13-2: A <code>main</code> function with hardcoded values to simulate user input and random number generation</span></p> <p>We’ve hardcoded the variable <code>simulated_user_specified_value</code> as 10 and the variable <code>simulated_random_number</code> as 7 for simplicity’s sake; in an actual program, we’d get the intensity number from the app frontend, and we’d use the <code>rand</code> crate to generate a random number, as we did in the Guessing Game example in Chapter 2. The <code>main</code> function calls a <code>generate_workout</code> function with the simulated input values.</p> <p>Now that we have the context, let’s get to the algorithm. The function <code>generate_workout</code> in Listing 13-3 contains the business logic of the app that we’re most concerned with in this example. The rest of the code changes in this example will be made to this function.</p> <p><span class="filename">Filename: src/main.rs</span></p> <pre data-language="rust">use std::thread;
use std::time::Duration;

fn simulated_expensive_calculation(intensity: u32) -&gt; u32 {
    println!("calculating slowly...");
    thread::sleep(Duration::from_secs(2));
    intensity
}

fn generate_workout(intensity: u32, random_number: u32) {
    if intensity &lt; 25 {
        println!(
            "Today, do {} pushups!",
            simulated_expensive_calculation(intensity)
        );
        println!(
            "Next, do {} situps!",
            simulated_expensive_calculation(intensity)
        );
    } else {
        if random_number == 3 {
            println!("Take a break today! Remember to stay hydrated!");
        } else {
            println!(
                "Today, run for {} minutes!",
                simulated_expensive_calculation(intensity)
            );
        }
    }
}

fn main() {
    let simulated_user_specified_value = 10;
    let simulated_random_number = 7;

    generate_workout(simulated_user_specified_value, simulated_random_number);
}
</pre> <p><span class="caption">Listing 13-3: The business logic that prints the workout plans based on the inputs and calls to the <code>simulated_expensive_calculation</code> function</span></p> <p>The code in Listing 13-3 has multiple calls to the slow calculation function. The first <code>if</code> block calls <code>simulated_expensive_calculation</code> twice, the <code>if</code> inside the outer <code>else</code> doesn’t call it at all, and the code inside the second <code>else</code> case calls it once.</p> <p>The desired behavior of the <code>generate_workout</code> function is to first check whether the user wants a low-intensity workout (indicated by a number less than 25) or a high-intensity workout (a number of 25 or greater).</p> <p>Low-intensity workout plans will recommend a number of push-ups and sit-ups based on the complex algorithm we’re simulating.</p> <p>If the user wants a high-intensity workout, there’s some additional logic: if the value of the random number generated by the app happens to be 3, the app will recommend a break and hydration. If not, the user will get a number of minutes of running based on the complex algorithm.</p> <p>This code works the way the business wants it to now, but let’s say the data science team decides that we need to make some changes to the way we call the <code>simulated_expensive_calculation</code> function in the future. To simplify the update when those changes happen, we want to refactor this code so it calls the <code>simulated_expensive_calculation</code> function only once. We also want to cut the place where we’re currently unnecessarily calling the function twice without adding any other calls to that function in the process. That is, we don’t want to call it if the result isn’t needed, and we still want to call it only once.</p> <h4 id="refactoring-using-functions">Refactoring Using Functions</h4> <p>We could restructure the workout program in many ways. First, we’ll try extracting the duplicated call to the <code>simulated_expensive_calculation</code> function into a variable, as shown in Listing 13-4.</p> <p><span class="filename">Filename: src/main.rs</span></p> <pre data-language="rust">use std::thread;
use std::time::Duration;

fn simulated_expensive_calculation(intensity: u32) -&gt; u32 {
    println!("calculating slowly...");
    thread::sleep(Duration::from_secs(2));
    intensity
}

fn generate_workout(intensity: u32, random_number: u32) {
    let expensive_result = simulated_expensive_calculation(intensity);

    if intensity &lt; 25 {
        println!("Today, do {} pushups!", expensive_result);
        println!("Next, do {} situps!", expensive_result);
    } else {
        if random_number == 3 {
            println!("Take a break today! Remember to stay hydrated!");
        } else {
            println!("Today, run for {} minutes!", expensive_result);
        }
    }
}

fn main() {
    let simulated_user_specified_value = 10;
    let simulated_random_number = 7;

    generate_workout(simulated_user_specified_value, simulated_random_number);
}
</pre> <p><span class="caption">Listing 13-4: Extracting the calls to <code>simulated_expensive_calculation</code> to one place and storing the result in the <code>expensive_result</code> variable</span></p> <p>This change unifies all the calls to <code>simulated_expensive_calculation</code> and solves the problem of the first <code>if</code> block unnecessarily calling the function twice. Unfortunately, we’re now calling this function and waiting for the result in all cases, which includes the inner <code>if</code> block that doesn’t use the result value at all.</p> <p>We want to refer to <code>simulated_expensive_calculation</code> only once in <code>generate_workout</code>, but defer the expensive calculation to only where we actually need the result. This is a use case for closures!</p> <h4 id="refactoring-with-closures-to-store-code">Refactoring with Closures to Store Code</h4> <p>Instead of always calling the <code>simulated_expensive_calculation</code> function before the <code>if</code> blocks, we can define a closure and store the <em>closure</em> in a variable rather than storing the result of the function call, as shown in Listing 13-5. We can actually move the whole body of <code>simulated_expensive_calculation</code> within the closure we’re introducing here.</p> <p><span class="filename">Filename: src/main.rs</span></p> <pre data-language="rust">use std::thread;
use std::time::Duration;

fn generate_workout(intensity: u32, random_number: u32) {
    let expensive_closure = |num| {
        println!("calculating slowly...");
        thread::sleep(Duration::from_secs(2));
        num
    };

    if intensity &lt; 25 {
        println!("Today, do {} pushups!", expensive_closure(intensity));
        println!("Next, do {} situps!", expensive_closure(intensity));
    } else {
        if random_number == 3 {
            println!("Take a break today! Remember to stay hydrated!");
        } else {
            println!(
                "Today, run for {} minutes!",
                expensive_closure(intensity)
            );
        }
    }
}

fn main() {
    let simulated_user_specified_value = 10;
    let simulated_random_number = 7;

    generate_workout(simulated_user_specified_value, simulated_random_number);
}
</pre> <p><span class="caption">Listing 13-5: Defining a closure and storing it in the <code>expensive_closure</code> variable</span></p> <p>The closure definition comes after the <code>=</code> to assign it to the variable <code>expensive_closure</code>. To define a closure, we start with a pair of vertical pipes (<code>|</code>), inside which we specify the parameters to the closure; this syntax was chosen because of its similarity to closure definitions in Smalltalk and Ruby. This closure has one parameter named <code>num</code>: if we had more than one parameter, we would separate them with commas, like <code>|param1, param2|</code>.</p> <p>After the parameters, we place curly brackets that hold the body of the closure—these are optional if the closure body is a single expression. The end of the closure, after the curly brackets, needs a semicolon to complete the <code>let</code> statement. The value returned from the last line in the closure body (<code>num</code>) will be the value returned from the closure when it’s called, because that line doesn’t end in a semicolon; just as in function bodies.</p> <p>Note that this <code>let</code> statement means <code>expensive_closure</code> contains the <em>definition</em> of an anonymous function, not the <em>resulting value</em> of calling the anonymous function. Recall that we’re using a closure because we want to define the code to call at one point, store that code, and call it at a later point; the code we want to call is now stored in <code>expensive_closure</code>.</p> <p>With the closure defined, we can change the code in the <code>if</code> blocks to call the closure to execute the code and get the resulting value. We call a closure like we do a function: we specify the variable name that holds the closure definition and follow it with parentheses containing the argument values we want to use, as shown in Listing 13-6.</p> <p><span class="filename">Filename: src/main.rs</span></p> <pre data-language="rust">use std::thread;
use std::time::Duration;

fn generate_workout(intensity: u32, random_number: u32) {
    let expensive_closure = |num| {
        println!("calculating slowly...");
        thread::sleep(Duration::from_secs(2));
        num
    };

    if intensity &lt; 25 {
        println!("Today, do {} pushups!", expensive_closure(intensity));
        println!("Next, do {} situps!", expensive_closure(intensity));
    } else {
        if random_number == 3 {
            println!("Take a break today! Remember to stay hydrated!");
        } else {
            println!(
                "Today, run for {} minutes!",
                expensive_closure(intensity)
            );
        }
    }
}

fn main() {
    let simulated_user_specified_value = 10;
    let simulated_random_number = 7;

    generate_workout(simulated_user_specified_value, simulated_random_number);
}
</pre> <p><span class="caption">Listing 13-6: Calling the <code>expensive_closure</code> we’ve defined</span></p> <p>Now how to perform the expensive calculation is defined in only one place, and we’re only executing that code where we need the results.</p> <p>However, we’ve reintroduced one of the problems from Listing 13-3: we’re still calling the closure twice in the first <code>if</code> block, which will call the expensive code twice and make the user wait twice as long as they need to. We could fix this problem by creating a variable local to that <code>if</code> block to hold the result of calling the closure, but closures provide us with another solution. We’ll talk about that solution in a bit. But first let’s talk about why there aren’t type annotations in the closure definition and the traits involved with closures.</p> <h3 id="closure-type-inference-and-annotation">Closure Type Inference and Annotation</h3> <p>Closures don’t require you to annotate the types of the parameters or the return value like <code>fn</code> functions do. Type annotations are required on functions because they’re part of an explicit interface exposed to your users. Defining this interface rigidly is important for ensuring that everyone agrees on what types of values a function uses and returns. But closures aren’t used in an exposed interface like this: they’re stored in variables and used without naming them and exposing them to users of our library.</p> <p>Closures are usually short and relevant only within a narrow context rather than in any arbitrary scenario. Within these limited contexts, the compiler is reliably able to infer the types of the parameters and the return type, similar to how it’s able to infer the types of most variables.</p> <p>Making programmers annotate the types in these small, anonymous functions would be annoying and largely redundant with the information the compiler already has available.</p> <p>As with variables, we can add type annotations if we want to increase explicitness and clarity at the cost of being more verbose than is strictly necessary. Annotating the types for the closure we defined in Listing 13-5 would look like the definition shown in Listing 13-7.</p> <p><span class="filename">Filename: src/main.rs</span></p> <pre data-language="rust">use std::thread;
use std::time::Duration;

fn generate_workout(intensity: u32, random_number: u32) {
    let expensive_closure = |num: u32| -&gt; u32 {
        println!("calculating slowly...");
        thread::sleep(Duration::from_secs(2));
        num
    };

    if intensity &lt; 25 {
        println!("Today, do {} pushups!", expensive_closure(intensity));
        println!("Next, do {} situps!", expensive_closure(intensity));
    } else {
        if random_number == 3 {
            println!("Take a break today! Remember to stay hydrated!");
        } else {
            println!(
                "Today, run for {} minutes!",
                expensive_closure(intensity)
            );
        }
    }
}

fn main() {
    let simulated_user_specified_value = 10;
    let simulated_random_number = 7;

    generate_workout(simulated_user_specified_value, simulated_random_number);
}
</pre> <p><span class="caption">Listing 13-7: Adding optional type annotations of the parameter and return value types in the closure</span></p> <p>With type annotations added, the syntax of closures looks more similar to the syntax of functions. The following is a vertical comparison of the syntax for the definition of a function that adds 1 to its parameter and a closure that has the same behavior. We’ve added some spaces to line up the relevant parts. This illustrates how closure syntax is similar to function syntax except for the use of pipes and the amount of syntax that is optional:</p> <pre data-language="rust">fn  add_one_v1   (x: u32) -&gt; u32 { x + 1 }
let add_one_v2 = |x: u32| -&gt; u32 { x + 1 };
let add_one_v3 = |x|             { x + 1 };
let add_one_v4 = |x|               x + 1  ;
</pre> <p>The first line shows a function definition, and the second line shows a fully annotated closure definition. The third line removes the type annotations from the closure definition, and the fourth line removes the brackets, which are optional because the closure body has only one expression. These are all valid definitions that will produce the same behavior when they’re called. Calling the closures is required for <code>add_one_v3</code> and <code>add_one_v4</code> to be able to compile because the types will be inferred from their usage.</p> <p>Closure definitions will have one concrete type inferred for each of their parameters and for their return value. For instance, Listing 13-8 shows the definition of a short closure that just returns the value it receives as a parameter. This closure isn’t very useful except for the purposes of this example. Note that we haven’t added any type annotations to the definition: if we then try to call the closure twice, using a <code>String</code> as an argument the first time and a <code>u32</code> the second time, we’ll get an error.</p> <p><span class="filename">Filename: src/main.rs</span></p> <pre data-language="rust">fn main() {
    let example_closure = |x| x;

    let s = example_closure(String::from("hello"));
    let n = example_closure(5);
}
</pre> <p><span class="caption">Listing 13-8: Attempting to call a closure whose types are inferred with two different types</span></p> <p>The compiler gives us this error:</p> <pre>$ cargo run
   Compiling closure-example v0.1.0 (file:///projects/closure-example)
error[E0308]: mismatched types
 --&gt; src/main.rs:5:29
  |
5 |     let n = example_closure(5);
  |                             ^
  |                             |
  |                             expected struct `String`, found integer
  |                             help: try using a conversion method: `5.to_string()`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: could not compile `closure-example`

To learn more, run the command again with --verbose.
</pre> <p>The first time we call <code>example_closure</code> with the <code>String</code> value, the compiler infers the type of <code>x</code> and the return type of the closure to be <code>String</code>. Those types are then locked into the closure in <code>example_closure</code>, and we get a type error if we try to use a different type with the same closure.</p> <h3 id="storing-closures-using-generic-parameters-and-the-fn-traits">Storing Closures Using Generic Parameters and the <code id="">Fn</code> Traits</h3> <p>Let’s return to our workout generation app. In Listing 13-6, our code was still calling the expensive calculation closure more times than it needed to. One option to solve this issue is to save the result of the expensive closure in a variable for reuse and use the variable in each place we need the result, instead of calling the closure again. However, this method could result in a lot of repeated code.</p> <p>Fortunately, another solution is available to us. We can create a struct that will hold the closure and the resulting value of calling the closure. The struct will execute the closure only if we need the resulting value, and it will cache the resulting value so the rest of our code doesn’t have to be responsible for saving and reusing the result. You may know this pattern as <em>memoization</em> or <em>lazy evaluation</em>.</p> <p>To make a struct that holds a closure, we need to specify the type of the closure, because a struct definition needs to know the types of each of its fields. Each closure instance has its own unique anonymous type: that is, even if two closures have the same signature, their types are still considered different. To define structs, enums, or function parameters that use closures, we use generics and trait bounds, as we discussed in Chapter 10.</p> <p>The <code>Fn</code> traits are provided by the standard library. All closures implement at least one of the traits: <code>Fn</code>, <code>FnMut</code>, or <code>FnOnce</code>. We’ll discuss the difference between these traits in the <a href="#capturing-the-environment-with-closures">“Capturing the Environment with Closures”</a> section; in this example, we can use the <code>Fn</code> trait.</p> <p>We add types to the <code>Fn</code> trait bound to represent the types of the parameters and return values the closures must have to match this trait bound. In this case, our closure has a parameter of type <code>u32</code> and returns a <code>u32</code>, so the trait bound we specify is <code>Fn(u32) -&gt; u32</code>.</p> <p>Listing 13-9 shows the definition of the <code>Cacher</code> struct that holds a closure and an optional result value.</p> <p><span class="filename">Filename: src/main.rs</span></p> <pre data-language="rust">struct Cacher&lt;T&gt;
where
    T: Fn(u32) -&gt; u32,
{
    calculation: T,
    value: Option&lt;u32&gt;,
}

fn main() {}
</pre> <p><span class="caption">Listing 13-9: Defining a <code>Cacher</code> struct that holds a closure in <code>calculation</code> and an optional result in <code>value</code></span></p> <p>The <code>Cacher</code> struct has a <code>calculation</code> field of the generic type <code>T</code>. The trait bounds on <code>T</code> specify that it’s a closure by using the <code>Fn</code> trait. Any closure we want to store in the <code>calculation</code> field must have one <code>u32</code> parameter (specified within the parentheses after <code>Fn</code>) and must return a <code>u32</code> (specified after the <code>-&gt;</code>).</p> <blockquote> <p>Note: Functions can implement all three of the <code>Fn</code> traits too. If what we want to do doesn’t require capturing a value from the environment, we can use a function rather than a closure where we need something that implements an <code>Fn</code> trait.</p> </blockquote> <p>The <code>value</code> field is of type <code>Option&lt;u32&gt;</code>. Before we execute the closure, <code>value</code> will be <code>None</code>. When code using a <code>Cacher</code> asks for the <em>result</em> of the closure, the <code>Cacher</code> will execute the closure at that time and store the result within a <code>Some</code> variant in the <code>value</code> field. Then if the code asks for the result of the closure again, instead of executing the closure again, the <code>Cacher</code> will return the result held in the <code>Some</code> variant.</p> <p>The logic around the <code>value</code> field we’ve just described is defined in Listing 13-10.</p> <p><span class="filename">Filename: src/main.rs</span></p> <pre data-language="rust">struct Cacher&lt;T&gt;
where
    T: Fn(u32) -&gt; u32,
{
    calculation: T,
    value: Option&lt;u32&gt;,
}

impl&lt;T&gt; Cacher&lt;T&gt;
where
    T: Fn(u32) -&gt; u32,
{
    fn new(calculation: T) -&gt; Cacher&lt;T&gt; {
        Cacher {
            calculation,
            value: None,
        }
    }

    fn value(&amp;mut self, arg: u32) -&gt; u32 {
        match self.value {
            Some(v) =&gt; v,
            None =&gt; {
                let v = (self.calculation)(arg);
                self.value = Some(v);
                v
            }
        }
    }
}

fn main() {}
</pre> <p><span class="caption">Listing 13-10: The caching logic of <code>Cacher</code></span></p> <p>We want <code>Cacher</code> to manage the struct fields’ values rather than letting the calling code potentially change the values in these fields directly, so these fields are private.</p> <p>The <code>Cacher::new</code> function takes a generic parameter <code>T</code>, which we’ve defined as having the same trait bound as the <code>Cacher</code> struct. Then <code>Cacher::new</code> returns a <code>Cacher</code> instance that holds the closure specified in the <code>calculation</code> field and a <code>None</code> value in the <code>value</code> field, because we haven’t executed the closure yet.</p> <p>When the calling code needs the result of evaluating the closure, instead of calling the closure directly, it will call the <code>value</code> method. This method checks whether we already have a resulting value in <code>self.value</code> in a <code>Some</code>; if we do, it returns the value within the <code>Some</code> without executing the closure again.</p> <p>If <code>self.value</code> is <code>None</code>, the code calls the closure stored in <code>self.calculation</code>, saves the result in <code>self.value</code> for future use, and returns the value as well.</p> <p>Listing 13-11 shows how we can use this <code>Cacher</code> struct in the function <code>generate_workout</code> from Listing 13-6.</p> <p><span class="filename">Filename: src/main.rs</span></p> <pre data-language="rust">use std::thread;
use std::time::Duration;

struct Cacher&lt;T&gt;
where
    T: Fn(u32) -&gt; u32,
{
    calculation: T,
    value: Option&lt;u32&gt;,
}

impl&lt;T&gt; Cacher&lt;T&gt;
where
    T: Fn(u32) -&gt; u32,
{
    fn new(calculation: T) -&gt; Cacher&lt;T&gt; {
        Cacher {
            calculation,
            value: None,
        }
    }

    fn value(&amp;mut self, arg: u32) -&gt; u32 {
        match self.value {
            Some(v) =&gt; v,
            None =&gt; {
                let v = (self.calculation)(arg);
                self.value = Some(v);
                v
            }
        }
    }
}

fn generate_workout(intensity: u32, random_number: u32) {
    let mut expensive_result = Cacher::new(|num| {
        println!("calculating slowly...");
        thread::sleep(Duration::from_secs(2));
        num
    });

    if intensity &lt; 25 {
        println!("Today, do {} pushups!", expensive_result.value(intensity));
        println!("Next, do {} situps!", expensive_result.value(intensity));
    } else {
        if random_number == 3 {
            println!("Take a break today! Remember to stay hydrated!");
        } else {
            println!(
                "Today, run for {} minutes!",
                expensive_result.value(intensity)
            );
        }
    }
}

fn main() {
    let simulated_user_specified_value = 10;
    let simulated_random_number = 7;

    generate_workout(simulated_user_specified_value, simulated_random_number);
}
</pre> <p><span class="caption">Listing 13-11: Using <code>Cacher</code> in the <code>generate_workout</code> function to abstract away the caching logic</span></p> <p>Instead of saving the closure in a variable directly, we save a new instance of <code>Cacher</code> that holds the closure. Then, in each place we want the result, we call the <code>value</code> method on the <code>Cacher</code> instance. We can call the <code>value</code> method as many times as we want, or not call it at all, and the expensive calculation will be run a maximum of once.</p> <p>Try running this program with the <code>main</code> function from Listing 13-2. Change the values in the <code>simulated_user_specified_value</code> and <code>simulated_random_number</code> variables to verify that in all the cases in the various <code>if</code> and <code>else</code> blocks, <code>calculating slowly...</code> appears only once and only when needed. The <code>Cacher</code> takes care of the logic necessary to ensure we aren’t calling the expensive calculation more than we need to so <code>generate_workout</code> can focus on the business logic.</p> <h3 id="limitations-of-the-cacher-implementation">Limitations of the <code id="">Cacher</code> Implementation</h3> <p>Caching values is a generally useful behavior that we might want to use in other parts of our code with different closures. However, there are two problems with the current implementation of <code>Cacher</code> that would make reusing it in different contexts difficult.</p> <p>The first problem is that a <code>Cacher</code> instance assumes it will always get the same value for the parameter <code>arg</code> to the <code>value</code> method. That is, this test of <code>Cacher</code> will fail:</p> <pre data-language="rust">struct Cacher&lt;T&gt;
where
    T: Fn(u32) -&gt; u32,
{
    calculation: T,
    value: Option&lt;u32&gt;,
}

impl&lt;T&gt; Cacher&lt;T&gt;
where
    T: Fn(u32) -&gt; u32,
{
    fn new(calculation: T) -&gt; Cacher&lt;T&gt; {
        Cacher {
            calculation,
            value: None,
        }
    }

    fn value(&amp;mut self, arg: u32) -&gt; u32 {
        match self.value {
            Some(v) =&gt; v,
            None =&gt; {
                let v = (self.calculation)(arg);
                self.value = Some(v);
                v
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn call_with_different_values() {
        let mut c = Cacher::new(|a| a);

        let v1 = c.value(1);
        let v2 = c.value(2);

        assert_eq!(v2, 2);
    }
}
</pre> <p>This test creates a new <code>Cacher</code> instance with a closure that returns the value passed into it. We call the <code>value</code> method on this <code>Cacher</code> instance with an <code>arg</code> value of 1 and then an <code>arg</code> value of 2, and we expect the call to <code>value</code> with the <code>arg</code> value of 2 to return 2.</p> <p>Run this test with the <code>Cacher</code> implementation in Listing 13-9 and Listing 13-10, and the test will fail on the <code>assert_eq!</code> with this message:</p> <pre>$ cargo test
   Compiling cacher v0.1.0 (file:///projects/cacher)
    Finished test [unoptimized + debuginfo] target(s) in 0.72s
     Running target/debug/deps/cacher-4116485fb32b3fff

running 1 test
test tests::call_with_different_values ... FAILED

failures:

---- tests::call_with_different_values stdout ----
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `2`', src/lib.rs:43:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::call_with_different_values

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

error: test failed, to rerun pass '--lib'
</pre> <p>The problem is that the first time we called <code>c.value</code> with 1, the <code>Cacher</code> instance saved <code>Some(1)</code> in <code>self.value</code>. Thereafter, no matter what we pass into the <code>value</code> method, it will always return 1.</p> <p>Try modifying <code>Cacher</code> to hold a hash map rather than a single value. The keys of the hash map will be the <code>arg</code> values that are passed in, and the values of the hash map will be the result of calling the closure on that key. Instead of looking at whether <code>self.value</code> directly has a <code>Some</code> or a <code>None</code> value, the <code>value</code> function will look up the <code>arg</code> in the hash map and return the value if it’s present. If it’s not present, the <code>Cacher</code> will call the closure and save the resulting value in the hash map associated with its <code>arg</code> value.</p> <p>The second problem with the current <code>Cacher</code> implementation is that it only accepts closures that take one parameter of type <code>u32</code> and return a <code>u32</code>. We might want to cache the results of closures that take a string slice and return <code>usize</code> values, for example. To fix this issue, try introducing more generic parameters to increase the flexibility of the <code>Cacher</code> functionality.</p> <h3 id="capturing-the-environment-with-closures">Capturing the Environment with Closures</h3> <p>In the workout generator example, we only used closures as inline anonymous functions. However, closures have an additional capability that functions don’t have: they can capture their environment and access variables from the scope in which they’re defined.</p> <p>Listing 13-12 has an example of a closure stored in the <code>equal_to_x</code> variable that uses the <code>x</code> variable from the closure’s surrounding environment.</p> <p><span class="filename">Filename: src/main.rs</span></p> <pre data-language="rust">fn main() {
    let x = 4;

    let equal_to_x = |z| z == x;

    let y = 4;

    assert!(equal_to_x(y));
}
</pre> <p><span class="caption">Listing 13-12: Example of a closure that refers to a variable in its enclosing scope</span></p> <p>Here, even though <code>x</code> is not one of the parameters of <code>equal_to_x</code>, the <code>equal_to_x</code> closure is allowed to use the <code>x</code> variable that’s defined in the same scope that <code>equal_to_x</code> is defined in.</p> <p>We can’t do the same with functions; if we try with the following example, our code won’t compile:</p> <p><span class="filename">Filename: src/main.rs</span></p> <pre data-language="rust">fn main() {
    let x = 4;

    fn equal_to_x(z: i32) -&gt; bool {
        z == x
    }

    let y = 4;

    assert!(equal_to_x(y));
}
</pre> <p>We get an error:</p> <pre>$ cargo run
   Compiling equal-to-x v0.1.0 (file:///projects/equal-to-x)
error[E0434]: can't capture dynamic environment in a fn item
 --&gt; src/main.rs:5:14
  |
5 |         z == x
  |              ^
  |
  = help: use the `|| { ... }` closure form instead

error: aborting due to previous error

For more information about this error, try `rustc --explain E0434`.
error: could not compile `equal-to-x`

To learn more, run the command again with --verbose.
</pre> <p>The compiler even reminds us that this only works with closures!</p> <p>When a closure captures a value from its environment, it uses memory to store the values for use in the closure body. This use of memory is overhead that we don’t want to pay in more common cases where we want to execute code that doesn’t capture its environment. Because functions are never allowed to capture their environment, defining and using functions will never incur this overhead.</p> <p>Closures can capture values from their environment in three ways, which directly map to the three ways a function can take a parameter: taking ownership, borrowing mutably, and borrowing immutably. These are encoded in the three <code>Fn</code> traits as follows:</p> <ul> <li>
<code>FnOnce</code> consumes the variables it captures from its enclosing scope, known as the closure’s <em>environment</em>. To consume the captured variables, the closure must take ownership of these variables and move them into the closure when it is defined. The <code>Once</code> part of the name represents the fact that the closure can’t take ownership of the same variables more than once, so it can be called only once.</li> <li>
<code>FnMut</code> can change the environment because it mutably borrows values.</li> <li>
<code>Fn</code> borrows values from the environment immutably.</li> </ul> <p>When you create a closure, Rust infers which trait to use based on how the closure uses the values from the environment. All closures implement <code>FnOnce</code> because they can all be called at least once. Closures that don’t move the captured variables also implement <code>FnMut</code>, and closures that don’t need mutable access to the captured variables also implement <code>Fn</code>. In Listing 13-12, the <code>equal_to_x</code> closure borrows <code>x</code> immutably (so <code>equal_to_x</code> has the <code>Fn</code> trait) because the body of the closure only needs to read the value in <code>x</code>.</p> <p>If you want to force the closure to take ownership of the values it uses in the environment, you can use the <code>move</code> keyword before the parameter list. This technique is mostly useful when passing a closure to a new thread to move the data so it’s owned by the new thread.</p> <blockquote> <p>Note: <code>move</code> closures may still implement <code>Fn</code> or <code>FnMut</code>, even though they capture variables by move. This is because the traits implemented by a closure type are determined by what the closure does with captured values, not how it captures them. The <code>move</code> keyword only specifies the latter.</p> </blockquote> <p>We’ll have more examples of <code>move</code> closures in Chapter 16 when we talk about concurrency. For now, here’s the code from Listing 13-12 with the <code>move</code> keyword added to the closure definition and using vectors instead of integers, because integers can be copied rather than moved; note that this code will not yet compile.</p> <p><span class="filename">Filename: src/main.rs</span></p> <pre data-language="rust">fn main() {
    let x = vec![1, 2, 3];

    let equal_to_x = move |z| z == x;

    println!("can't use x here: {:?}", x);

    let y = vec![1, 2, 3];

    assert!(equal_to_x(y));
}
</pre> <p>We receive the following error:</p> <pre>$ cargo run
   Compiling equal-to-x v0.1.0 (file:///projects/equal-to-x)
error[E0382]: borrow of moved value: `x`
 --&gt; src/main.rs:6:40
  |
2 |     let x = vec![1, 2, 3];
  |         - move occurs because `x` has type `Vec&lt;i32&gt;`, which does not implement the `Copy` trait
3 | 
4 |     let equal_to_x = move |z| z == x;
  |                      --------      - variable moved due to use in closure
  |                      |
  |                      value moved into closure here
5 | 
6 |     println!("can't use x here: {:?}", x);
  |                                        ^ value borrowed here after move

error: aborting due to previous error

For more information about this error, try `rustc --explain E0382`.
error: could not compile `equal-to-x`

To learn more, run the command again with --verbose.
</pre> <p>The <code>x</code> value is moved into the closure when the closure is defined, because we added the <code>move</code> keyword. The closure then has ownership of <code>x</code>, and <code>main</code> isn’t allowed to use <code>x</code> anymore in the <code>println!</code> statement. Removing <code>println!</code> will fix this example.</p> <p>Most of the time when specifying one of the <code>Fn</code> trait bounds, you can start with <code>Fn</code> and the compiler will tell you if you need <code>FnMut</code> or <code>FnOnce</code> based on what happens in the closure body.</p> <p>To illustrate situations where closures that can capture their environment are useful as function parameters, let’s move on to our next topic: iterators.</p><div class="_attribution">
  <p class="_attribution-p">
    <a href="https://doc.rust-lang.org/book/ch13-01-closures.html" class="_attribution-link" target="_blank">https://doc.rust-lang.org/book/ch13-01-closures.html</a>
  </p>
</div>
